libraries/lib-time-and-pitch/FormantShifterLogger.cpp

```diff
@@ -0,0 +1,122 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  FormantShifterLogger.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "FormantShifterLogger.h"
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+namespace
+{
+const std::string logDir { CMAKE_CURRENT_SOURCE_DIR };
```

Comment:
I believe you only use it for tests. It's way too intrusive, TBH

Reply:
It's convenient. What problem do you see with it?

Reply:
Provided that you need a custom config in Audacity itself, why not use the "Setting"?

Reply:
> It's convenient. What problem do you see with it?

It is a hack that hardcodes some random path into the build for no reason.

Reply:
> Provided that you need a custom config in Audacity itself, why not use the "Setting"?

You mean as defined in `Prefs.h`? I'd still need a write mechanism.

Reply:
> I'd still need a write mechanism.

Can you explain please?

Reply:
Sure: if e.g. a method like `std::optional<int> GetFftSizeOverride()` finds `overrideFftSizeExponent.txt` in `logDir` that has a valid int value, it will use this rather than the production value. I can hear the result change on the fly by editing that file during playback. Even if I used the prefs, I'd need this functionality, that's not exposed to the user. (And in this regard it's actually not bad that `logDir` has some random string in the production build, as it makes it near impossible for the user to inadvertently have that file at that path and get a non-production parameter.)

Reply:
Well, you've just made it worse. During the playback, do you randomly try to open a file? I wouldn't call that a good idea at all. Open/read/close are syscalls. They are not cheap, and they can cause havoc in some scenarios. Do we want such an approach to support debug scenarios tied strictly to one machine? I say we don't.

> And in this regard it's actually not bad that logDir has some random string in the production build

Linux package maintainers will disagree. You make the build result depend on the file system structure of the computer used to build it. 

---

I'm very much confused by this approach. Probably, it doesn't cause any noticeable difference for you, but I am very much against it.

Please create a separate config for the pitch-shifter. If you want to stick to reading files for now, please make access to the files conditionally compiled. I also prefer you use an executable/data path as a root for those files.

In the future, if we want to expose these settings to users, it would be a very quick and easy operation.

Reply:
> You make the build result depend on the file system structure of the computer used to build it.

I don't understand how does that matters: it's intended for development and functional only for local builds. But I don't mind using an app-data directory. I was lazy but it has advantages.

> Open/read/close are syscalls. They are not cheap, and they can cause havoc in some scenarios.

I didn't have this in mind. I want to reduce conditional compilation to a maximum to avoid code becoming outdated, but if it's only just around the file opening that's completely fine.

Reply:
>  it's intended for development and functional only for local builds

You push it to production builds, and that matters.

> I want to reduce conditional compilation to a maximum

You push the code only needed for the feature tuning into the production build. Expensive code.